waveorder/stokes.py

```diff
@@ -0,0 +1,338 @@
+import numpy as np
+
+"""
+This module collects Stokes- and Mueller-related calculations. 
+
+The functions are organized into three groups:
+
+1) A forward function group: 
+
+A = A_matrix(swing, scheme="5-State")
+s0, s1, s3, s3 = s0123_CPL_after_ADR(ret, ori, tra, dop)
+s0, s1, s2 = s012_CPL_after_AR(ret, ori, dop)
+
+2) An inverse function group:
+
+inverse_s0123_CPL_after_ADR(s0, s1, s2, s3)
+inverse s012_CPL_after_AR(s0, s1, s2)
+AR_mueller_from_CPL_projection(s0, s1, s2, s3)
+
+3) A convenience function group:
+
+M = inv_AR_mueller_from_CPL_projection(s0, s1, s2, s3)
+y = matmul(A, x)
+
+----------
+
+Usage: 
+
+All functions (except A_matrix) are intended to be used with ND-arrays with
+Stokes- or Mueller-indices as the first axes. 
+
+For example, the following usage modes of s0123_CPL_after_ADR are valid:
+
+>>> s0123_CPL_after_ADR(1, 1, 1, 1)
+
+>>> ret = np.ones((2,3,4,5))
+>>> ori = np.ones((2,3,4,5))
+>>> tra = np.ones((2,3,4,5))
+>>> dop = np.ones((2,3,4,5))
+>>> s0123_CPL_after_ADR(ret, ori, tra, dop)
+
+>>> ADR_params = np.ones((4,2,3,4,5)) # first axis contains the Stokes indices
+>>> s0123_CPL_after_ADR(*ADR_params) # * expands along the first axis
+
+"""
+
+
+# Forward function group
+
+
+def A_matrix(swing, scheme="5-State"):
+    """
+    Calculate the polarimeter system matrix for a swing and calibration scheme.
+
+    Parameters
+    ----------
+    swing : float
+        Result is periodic on the integers, e.g. A_matrix(0.1) = A_matrix(1.1)
+    scheme : str, "4-State" or "5-State"
```

Comment:
Another syntactical benefit of type hints is that `str, "4-State" or "5-State"` can be `Literal["4-State", "5-State"]` instead.

Reply:
This is where I got confused. Should I:

1) Change `scheme : str, "4-State" or "5-State"` to `scheme : Literal["4-State", "5-State"]`. Can a type checker use the docstring?

2) Use type annotations in the function signature? Does this mean I'll need to maintain both the docstring and the type annotation, or is there a shortcut?

Reply:
> Can a type checker use the docstring?

Not that I'm aware of.

> Use type annotations in the function signature? Does this mean I'll need to maintain both the docstring and the type annotation, or is there a shortcut?

In creation time docstrings can be polulated from the signature, but for the plugin I'm using they have to be manually synced in later editing. There might be other tools out there though.